esp/platform/pluginloader.hpp

```diff
@@ -0,0 +1,184 @@
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _PluginLoader_HPP_
+#define _PluginLoader_HPP_
+
+#include "jlog.hpp"
+#include "jmutex.hpp"
+#include "jptree.hpp"
+#include <map>
+#include <string>
+
+/**
+ * TPluginLoader encapsulates the logic used to obtain an entry-point function pointer from a
+ * shared library. The logic involved is not complex, but it is repetitive.
+ *
+ * The shared library name may be configured dynamically or known in advance. Given a
+ * configuration property tree and an XPath, the name will be extracted from the property tree.
+ * Given only a default name, the default will be used. Given all three, a configured value will
+ * be used if one exists and the default will be used otherwise.
+ *
+ * The entry-point function name may be configured dynamically or known in advance. Given a
+ * configuration property tree and an XPath, the name will be extracted from the property tree.
+ * Given only a default name, the default will be used. Given all three, a configured value will
+ * be used it one exists and the default will be used otherwise.
+ *
+ * The template parameter is an entry point function signature. This suggests a limitation of
+ * one kind of plugin per instance.
+ */
+template <typename entry_point_t>
+class TPluginLoader
+{
+public:
+    template <typename instance_t>
+    using Creator = std::function<instance_t*(entry_point_t entryPoint)>;
+
+    TPluginLoader(const char* libraryDefault, const char* entryPointDefault, const char* libraryXPath = nullptr, const char* entryPointXPath = nullptr)
+        : m_libraryXPath(libraryXPath)
+        , m_libraryDefault(libraryDefault)
+        , m_entryPointXPath(entryPointXPath)
+        , m_entryPointDefault(entryPointDefault)
+    {
+        StringBuffer missing;
+        if (m_libraryXPath.isEmpty() && m_libraryDefault.isEmpty())
+        {
+            if (!missing.isEmpty())
+                missing.append(" and ");
+            missing.append("library name");
+        }
+        if (m_entryPointXPath.isEmpty() && m_entryPointDefault.isEmpty())
+        {
+            if (!missing.isEmpty())
+                missing.append(" and ");
+            missing.append("entry-point function name");
+        }
+        if (!missing.isEmpty())
+            ERRLOG("bad plugin loading configuration - missing %s", missing.str());
+    }
+
+    template <typename instance_t>
+    instance_t* create(const IPTree& configuration, Creator<instance_t> creator)
```

Comment:
I don't see the significance of having the "Creator" functor, which is simply a pass through of the entry_point_t, and entry_point_t is already known as soon as you declare m_loader. My question basically is, why do we need the functor?

Reply:
With this initial use case, the entry point takes no parameters and would be easy to call directly. Not all plugin entry points take no parameters. The functor hides entry point signature differences.

Reply:
In this case it's fine to keep the functor for future extension.